### DISCUSSION TITLE
Spark EMR - training splink on chunks or single, very large datasets?

### DISCUSSION THREAD

**Discussion by jkginfinite** [moj-analytical-services/splink#1552] [Q&A]
I have a full dataset of around 160million rows. We've got some good blocking rules but it seems to fail on datasets larger than 10million rows.

I am wondering if its typical Splink practice for a huge dataset (far exceeding 10 million rows) to broken apart and trained in chunks as opposed to training the whole dataset at once.

Is chunking AKA batch processing typical with Splink (much like image models with neural networks) the more common way to train splink? Or  is doing 40... 50 100 million rows common all at once? 

(we have 10 columns in this dataset)

**Comment by RobinL** [comment]
That's kind of what salting is supposed to do, see here:
https://moj-analytical-services.github.io/splink/topic_guides/performance/salting.html?h=salting

The idea is that we want spark to chunk the work up into as many tasks as possible.  

Note though that salting only applies to predictions, not to training.

There's a bit more info here:
https://github.com/moj-analytical-services/splink/issues/527
https://github.com/moj-analytical-services/splink/discussions/254

For training, note that you can use tight blocking rules.  See [here](https://moj-analytical-services.github.io/splink/topic_guides/blocking/model_training.html):
> Unlike [Prediction Rules](https://moj-analytical-services.github.io/splink/topic_guides/blocking/predictions.html), it does not matter if Training Rules excludes some true matches - it just needs to generate examples of matches and non-matches.

Before resorting to salting, if you haven't already, you could try the suggestions [here](https://moj-analytical-services.github.io/splink/topic_guides/performance/optimising_spark.html?h=optimising)

**Reply by jkginfinite** [reply]
is this available in the latest version of Splink? 

Does this work by making predictions in  partitions (salt, or whatever we call it) and then we can push those "salts" to wherever they need to be sent?

What would the code look like to do this? I can see the code for putting the salt in the blocking rules, but I didn't see anything different in the output predictions. 

I am wondering how the output might change such that I could push my predictions in stages and avoid memory running out?
